### PR TITLE
feat: remake diagnostic graph packages

### PR DESCRIPTION
**Pull request by isamu-takagi**
## Description

This is a continuation of the diagnostic graph proposed in https://github.com/autowarefoundation/autoware.universe/issues/3829. The previous diagnostic graph sent static data every frame, so I improved efficiency by separating static and dynamic data. Also, I added diagnostic graph tools to improve usability.

This PR contains updates for the following packages

- diagnostic_graph_aggregator: The core package of diagnostic graph.
- diagnostic_graph_utils: Tools and utility classes to subscribe diagnostic graph.
- hazard_status_converter: A tool to convert diagnostic graph to hazard status for backward compatibilty.
- system_diagnostic_monitor: Configuration files and scripts to integrate diagnostic graph into Autoware.
- rqt_diagnostic_graph_monitor: Visualization tools for diagnostic graph.

Screenshot from 2024-03-31 04-59-43

## Related links

This PR depends on https://github.com/tier4/tier4_autoware_msgs/pull/118.

## Tests performed

Please check results with the rqt monitor unless otherwise specified. Launch it with the following command.
```bash
ros2 run rqt_diagnostic_graph_monitor rqt_diagnostic_graph_monitor
```
0. Merge dependent PR.
    - https://github.com/tier4/tier4_autoware_msgs/pull/118.

1. Check examples
```bash
ros2 launch diagnostic_graph_aggregator example-main.launch.xml
```
```bash
ros2 launch diagnostic_graph_aggregator example-exit.launch.xml
```

2. Check config [tree view tool](https://github.com/autowarefoundation/autoware.universe/blob/5f0a6e9bb152f998c42e246002a3d6bc56a6f807/system/diagnostic_graph_aggregator/doc/tool/tree.md) for examples (The result will be displayed in the terminal).

```bash
ros2 run diagnostic_graph_aggregator tree '$(find-pkg-share diagnostic_graph_aggregator)/example/graph/naub.yaml'
```
```bash
ros2 run diagnostic_graph_aggregator tree '$(find-pkg-share diagnostic_graph_aggregator)/example/graph/edit.yaml'
```

3. Check [converter tool](https://github.com/autowarefoundation/autoware.universe/blob/5f0a6e9bb152f998c42e246002a3d6bc56a6f807/system/diagnostic_graph_utils/doc/node/converter.md) (check result by rqt_runtime_monitor).

```bash
Terminal 1: ros2 launch diagnostic_graph_aggregator example-main.launch.xml
Terminal 2: ros2 run diagnostic_graph_utils converter
Terminal 3: ros2 run rqt_runtime_monitor rqt_runtime_monitor --ros-args -r diagnostics:=diagnostics_array
```

3. Check [dump tool](https://github.com/autowarefoundation/autoware.universe/blob/5f0a6e9bb152f998c42e246002a3d6bc56a6f807/system/diagnostic_graph_utils/doc/node/dump.md) (The result will be displayed in the terminal).

```bash
Terminal 1: ros2 launch diagnostic_graph_aggregator example-main.launch.xml
Terminal 2: ros2 run diagnostic_graph_utils dump
```

4. Check planning simulator. Check if all status is OK by initializing pose and set route.

```bash
Terminal 1: (Launch planning simulator)
Terminal 2: ros2 launch system_diagnostic_monitor system_diagnostic_monitor.launch.xml graph_file:='$(find-pkg-share system_diagnostic_monitor)/config/autoware-psim.yaml'
```

5. Check hazard_status_converter (Check result by ros2 topic echo).

```bash
Terminal 1: (Launch planning simulator)
Terminal 2: ros2 launch system_diagnostic_monitor system_diagnostic_monitor.launch.xml graph_file:='$(find-pkg-share system_diagnostic_monitor)/config/autoware-psim.yaml'
Terminal 3: ros2 launch hazard_status_converter hazard_status_converter.launch.xml
Tarminal 4: ros2 topic echo /system/emergency/hazard_status
```

## Interface changes

- Add `/diagnostics_graph/struct`
- Add `/diagnostics_graph/status`
- Add `/diagnostics_graph/unknowns`

## Effects on system behavior

None at this time because not launched by default.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by TomohitoAndo** [comment]
@takagi
Could you resolve the spell-check-partial CI error?
https://github.com/autowarefoundation/autoware.universe/actions/runs/8641357471/job/23690752744?pr=6715

**Comment by isamu-takagi** [comment]
@TomohitoAndo Add dictionary. Could you review this PR? https://github.com/tier4/autoware-spell-check-dict/pull/712

**Comment by TomohitoAndo** [comment]
@isamu-takagi 
How can I use the remapping for the intermediate path?
For example, using `warn-to-ok` like following was acceptable previously, but this causes error now.

```
units:
  - path: /functions/obstacle_detection-error
    type: warn-to-ok
    list:
    - { type: link, link: /functions/obstacle_detection }

  - path: /functions/obstacle_detection
    type: or
    list:
      - { type: link, link: /sensing/lidars/front }
      - { type: link, link: /sensing/radars/front }
```

The error is as follows.
```
[aggregator-1] terminate called after throwing an instance of 'std::out_of_range'
[aggregator-1]   what():  _Map_base::at
[ERROR] [aggregator-1]: process has died [pid 2085983, exit code -6, cmd '/home/tomohitoando/workspace/autoware/install/diagnostic_graph_aggregator/lib/diagnostic_graph_aggregator/aggregator --ros-args -r __node:=diagnostic_graph_aggregator_self --params-file /home/tomohitoando/workspace/autoware/install/diagnostic_graph_aggregator/share/diagnostic_graph_aggregator/config/default.param.yaml --params-file /tmp/launch_params_ul9ov2z_'].
```

**Comment by isamu-takagi** [comment]
@TomohitoAndo Nodes that take only one child are now specified by `item`. Could you try it instead of `list`?

**Comment by TomohitoAndo** [comment]
@isamu-takagi 
I changed the file as follows, but the same error occurs :thinking: 
```
units:
  - path: /functions/obstacle_detection-error
    type: warn-to-ok
    item: { type: link, link: /functions/obstacle_detection }

  - path: /functions/pose_estimation
    type: and
    list:
      - { type: link, link: /sensing/lidars/top }

  - path: /functions/obstacle_detection
    type: or
    list:
      - { type: link, link: /sensing/lidars/front }
      - { type: link, link: /sensing/radars/front }

  - path: /sensing/lidars/top
    type: diag
    node: lidar_driver_top
    name: status

  - path: /sensing/lidars/front
    type: diag
    node: lidar_driver_front
    name: status

  - path: /sensing/radars/front
    type: diag
    node: radar_driver_front
    name: status
```

```
[aggregator-1] terminate called after throwing an instance of 'std::out_of_range'
[aggregator-1]   what():  _Map_base::at
[ERROR] [aggregator-1]: process has died [pid 1291015, exit code -6, cmd '/home/tomohitoando/workspace/autoware/install/diagnostic_graph_aggregator/lib/diagnostic_graph_aggregator/aggregator --ros-args -r __node:=diagnostic_graph_aggregator_self --params-file /home/tomohitoando/workspace/autoware/install/diagnostic_graph_aggregator/share/diagnostic_graph_aggregator/config/default.param.yaml --params-file /tmp/launch_params_ot04djcf'].
```

**Comment by TomohitoAndo** [comment]
@isamu-takagi 
I confirmed the error disappeared. :+1:

**Review by TomohitoAndo — APPROVED** [review]
As for `diagnostic_graph_aggregator`, `diagnostic_graph_utils` and `rqt_diagnostic_graph_monitor`, LGTM!

**Review by TetsuKawa — APPROVED** [review]
`hazard_ststaus_converter`, `system_diagnostic_monitor` LGTM.

File: system/diagnostic_graph_aggregator/README.md [hunk 1]
```diff
@@ -3,20 +3,25 @@
...
 
-## Diagnostics graph message
+## Diagnostic graph structures
 
-The diagnostics graph that this node outputs is a combination of diagnostic status and connections between them.
-This graph consists of an array of diagnostic nodes, and each node has a status and links.
-This link contains an index indicating the position of the node in the graph.
-Therefore, the graph can be reconstructed from the array of nodes using links.
-The following is an example of a message representing the graph in the overview section.
+The diagnostic graph is actually a set of fault tree analysis (FTA) for for each operation mode of Autoware.
```
Comment (TetsuKawa) [inline system/diagnostic_graph_aggregator/README.md:12]: A `for` is unnecessary.

Comment (isamu-takagi) [reply system/diagnostic_graph_aggregator/README.md]: Thank you. Fixed it https://github.com/autowarefoundation/autoware.universe/pull/6715/commits/cdaa322c9470d7fce62b64896326b4f028c401be.

**Comment by Owen-Liuyuxuan** [comment]
I hope we can have a feature similar to the `isInNoFaultCondition` in `system_error_monitor` so that the node can refrain from printing warnings and errors when the vehicle is not initialized (or only prints one warning every minute).

**Comment by isamu-takagi** [comment]
@Owen-Liuyuxuan 
diagnostic_graph_monitor does not need that feature since it diagnoses multiple modes separately. During initialization, it focuses on the diagnostics for STOP mode, and only includes diagnostics that require initialization for AUTO mode (also any other modes that require it).

**Comment by Owen-Liuyuxuan** [comment]
@isamu-takagi 
So should I assume that there is a bug somewhere in the system, when I see unexpected warning messages in the terminal during the starting stage of PSim, when initialized all those warnings will disappear?

**Comment by isamu-takagi** [comment]
@Owen-Liuyuxuan 
This is not necessarily the case. Each node outputs logs regardless of whether it is initializing or not. This is the same for system_error_monitor. `isInNoFaultCondition` seems to only affect logs output by system_error_monitor, but diagnostic_graph_aggregator does not output logs, so it should not be related.

**Pull request merged**